src/mds/MDSRank.cc

```diff
@@ -901,9 +901,29 @@ Session *MDSRank::get_session(Message *m)
 {
   Session *session = static_cast<Session *>(m->get_connection()->get_priv());
   if (session) {
+    session->put(); // do not carry ref
     dout(20) << "get_session have " << session << " " << session->info.inst
 	     << " state " << session->get_state_name() << dendl;
-    session->put();  // not carry ref
+    // Check if we've imported an open session since (new sessions start closed)
+    if (session->is_closed()) {
+      Session *imported_session = sessionmap.get_session(session->info.inst.name);
+      if (imported_session && imported_session != session) {
+        dout(10) << __func__ << " replacing connection bootstrap session " << session << " with imported session " << imported_session << dendl;
+        imported_session->info.auth_name = session->info.auth_name;
+        //assert(session->info.auth_name == imported_session->info.auth_name);
+        assert(session->info.inst == imported_session->info.inst);
+        imported_session->connection = session->connection;
+        // send out any queued messages
+        while (!session->preopen_out_queue.empty()) {
+          imported_session->connection->send_message(session->preopen_out_queue.front());
+          session->preopen_out_queue.pop_front();
```

Comment:
Eek, I haven't seen this preopen_out_queue before; is it protected by the global MDS lock? Otherwise I'd be worried about the locking rules here, not to mention a formerly-cheap lookup function turning into something that might get hit by message throttles...(Not that I think we have any on outgoing client messages right now.)

Reply:
That was added in: db976663a50ce7ab2448c42cb0c6bfd4f58d25a5

Reply:
> is it protected by the global MDS lock?

Yes.

In the path we're in, we already have it locked. I don't see any locking issues.